test/e2e/dns_autoscaling.go

```diff
@@ -96,54 +91,78 @@ var _ = framework.KubeDescribe("DNS horizontal autoscaling", func() {
+		Expect(updateDNSScalingConfigMap(c, packDNSScalingConfigMap(DNSParams_2.data))).NotTo(HaveOccurred())
 		By("Wait for kube-dns scaled to expected number")
-		ExpectNoError(waitForDNSReplicasSatisfied(c, int(math.Ceil(float64(nodeCount)/2.0)), DNSdefaultTimeout))
+		getExpectReplicasLinear := getExpectReplicasFuncLinear(c, DNSParams_2.nodesPerReplica)
+		Expect(waitForDNSReplicasSatisfied(c, getExpectReplicasLinear, DNSdefaultTimeout)).NotTo(HaveOccurred())
 
 		By("--- Scenario: should re-create scaling parameters with default value when parameters got deleted ---")
 		By("Delete the ConfigMap for autoscaler")
 		err := deleteDNSScalingConfigMap(c)
-		ExpectNoError(err)
+		Expect(err).NotTo(HaveOccurred())
 
 		By("Wait for the ConfigMap got re-created")
 		configMap, err := waitForDNSConfigMapCreated(c, DNSdefaultTimeout)
-		ExpectNoError(err)
+		Expect(err).NotTo(HaveOccurred())
 
 		By("Check the new created ConfigMap got the same data as we have")
 		Expect(reflect.DeepEqual(previousParams, configMap.Data)).To(Equal(true))
 
 		By("Replace the dns autoscaling parameters with the second testing parameters")
-		ExpectNoError(updateDNSScalingConfigMap(c, packDNSScalingConfigMap(DNSParams_2)))
+		Expect(updateDNSScalingConfigMap(c, packDNSScalingConfigMap(DNSParams_1.data))).NotTo(HaveOccurred())
 		By("Wait for kube-dns scaled to expected number")
-		ExpectNoError(waitForDNSReplicasSatisfied(c, int(math.Ceil(float64(nodeCount)/2.0)), DNSdefaultTimeout))
+		getExpectReplicasLinear = getExpectReplicasFuncLinear(c, DNSParams_1.nodesPerReplica)
+		Expect(waitForDNSReplicasSatisfied(c, getExpectReplicasLinear, DNSdefaultTimeout)).NotTo(HaveOccurred())
 
 		By("--- Scenario: should recover after autoscaler pod got deleted ---")
 		By("Delete the autoscaler pod for kube-dns")
-		ExpectNoError(deleteDNSAutoscalerPod(c))
+		Expect(deleteDNSAutoscalerPod(c)).NotTo(HaveOccurred())
 
 		By("Replace the dns autoscaling parameters with the third testing parameters")
-		ExpectNoError(updateDNSScalingConfigMap(c, packDNSScalingConfigMap(DNSParams_3)))
+		Expect(updateDNSScalingConfigMap(c, packDNSScalingConfigMap(DNSParams_3.data))).NotTo(HaveOccurred())
 		By("Wait for kube-dns scaled to expected number")
-		ExpectNoError(waitForDNSReplicasSatisfied(c, int(math.Ceil(float64(nodeCount)/3.0)), DNSdefaultTimeout))
+		getExpectReplicasLinear = getExpectReplicasFuncLinear(c, DNSParams_3.nodesPerReplica)
+		Expect(waitForDNSReplicasSatisfied(c, getExpectReplicasLinear, DNSdefaultTimeout)).NotTo(HaveOccurred())
 	})
 })
 
+type DNSParamsLinear struct {
+	data            map[string]string
+	nodesPerReplica float64
+}
+
+type getExpectReplicasFunc func(c clientset.Interface) int
+
+func getExpectReplicasFuncLinear(c clientset.Interface, nodesPerReplica float64) getExpectReplicasFunc {
+	return func(c clientset.Interface) int {
+		nodeCount := numOfSchedulableNodes(c)
+		return int(math.Ceil(float64(nodeCount) / nodesPerReplica))
```

Comment:
I considered the unit tests already cover that part. But yes, it would be better if we test that as well.